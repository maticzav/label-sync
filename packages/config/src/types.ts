import { z } from 'zod'

/**
 * Label represents the central unit of LabelSync. Each label
 * can have multiple siblings that are meaningfully related to
 * a label itself, multiple hooks that trigger different actions.
 */
export const LSCLabel = z.object({
  color: z.string(),
  description: z.string().optional(),
  siblings: z.array(z.string()).optional(),
  alias: z.array(z.string()).optional(),
  scope: z.array(z.string()).optional(),
})

export type LSCLabel = z.infer<typeof LSCLabel>

export const LSCLabelName = z.string()
export type LSCLabelName = z.infer<typeof LSCLabelName>

/**
 * Repository configuration for how LabelSync should sync it.
 */
const LSCRepositoryConfiguration = z.object({
  removeUnconfiguredLabels: z.boolean().optional(),
})
export interface LSCRepositoryConfiguration extends z.infer<typeof LSCRepositoryConfiguration> {}

/**
 * Repository represents a single Github repository.
 * When configured as `strict` it will delete any surplus of labels
 * in the repository.
 */
export const LSCRepository = z.object({
  config: LSCRepositoryConfiguration.optional(),
  labels: z.record(LSCLabelName, LSCLabel),
})
export interface LSCRepository extends z.infer<typeof LSCRepository> {}

export const LSCRepositoryName = z.string()
export type LSCRepositoryName = z.infer<typeof LSCRepositoryName>

/**
 * Configuration represents an entire configuration for all
 * LabelSync tools that an organisation is using.
 */
export const LSCConfiguration = z.object({
  repos: z.record(LSCRepositoryName, LSCRepository),
})
export interface LSCConfiguration extends z.infer<typeof LSCConfiguration> {}
