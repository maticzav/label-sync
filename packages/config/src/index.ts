export { LSCConfiguration, LSCRepositoryConfiguration, LSCRepository, LSCLabel } from './types'
export {
  parseConfig,
  isConfigRepo,
  getPhysicalRepositories,
  normalizeColor,
  getLSConfigRepoName,
  LS_CONFIG_PATH,
} from './parse'
