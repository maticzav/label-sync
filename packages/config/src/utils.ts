export type Dict<T> = { [key: string]: T }

/**
 * Maps entries in an object.
 * @param m
 * @param fn
 */
export function mapEntries<T, V>(m: Dict<T>, fn: (v: T, key: string) => V): Dict<V> {
  const mapped = Object.keys(m).map((key) => {
    return [key, fn(m[key], key)]
  })
  return Object.fromEntries(mapped)
}

/**
 * Maps keys of an object.
 * @param m
 * @param fn
 */
export function mapKeys<T>(m: Dict<T>, fn: (key: string, v: T) => string): Dict<T> {
  const mapped = Object.keys(m).map((key) => {
    return [fn(key, m[key]), m[key]]
  })
  return Object.fromEntries(mapped)
}
