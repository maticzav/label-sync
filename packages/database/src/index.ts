export { InstallationsSource } from './sources/installations'
export { TasksSource } from './sources/tasks'
export { Installation } from '@prisma/client'
