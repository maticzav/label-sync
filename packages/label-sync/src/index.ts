export { repo, label } from './generator'
export { labelsync } from './make'
export * from './presets'
