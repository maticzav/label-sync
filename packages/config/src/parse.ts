import yaml from 'js-yaml'
import ml from 'multilines'
import * as os from 'os'

import { LSCConfiguration, LSCLabel, LSCRepository } from './types'
import { mapEntries, mapKeys } from './utils'

/* Constants */

const NUMBER_OF_FREE_TIERS = 5

/**
 * Configuration repository is the repository which LabelSync
 * uses to determine the configuration of the service.
 */
export const getLSConfigRepoName = (account: string) => `${account.toLowerCase()}-labelsync`

/**
 * Configuration file path determines the path of the file in the repositoy
 * that we use to gather configuration information from.
 *
 * It should always be in YAML format.
 */
export const LS_CONFIG_PATH = 'labelsync.yml'

/**
 * Parses a string into a configuration object.
 */
export function parseConfig({
  input,
  isPro,
}: {
  input: string
  isPro: boolean
}): { ok: true; config: LSCConfiguration } | { ok: false; error: string } {
  const rawConfig = yaml.load(input, {})
  const parsedConfig = LSCConfiguration.safeParse(rawConfig)

  if (!parsedConfig.success) {
    return { ok: false, error: parsedConfig.error.message }
  }

  const normalizedConfig = normalizeConfig(parsedConfig.data)
  const validatedConfig = validateConfiguration(normalizedConfig)

  if (!validatedConfig.ok) {
    return { ok: false, error: validatedConfig.error }
  }

  const acceptable = checkPurchase({
    config: validatedConfig.config,
    isPro,
  })

  if (!acceptable.ok) {
    return { ok: false, error: acceptable.error }
  }

  return { ok: true, config: validatedConfig.config }
}

/**
 * Makes sure that the config respects plan limitations..
 */
function checkPurchase({
  config,
  isPro,
}: {
  config: LSCConfiguration
  isPro: boolean
}): { ok: true } | { ok: false; error: string } {
  const numberOfConfiguredRepos = Object.keys(config.repos).length

  if (isPro) {
    return { ok: true }
  }

  /**
   * Report too many configurations.
   */
  if (numberOfConfiguredRepos >= NUMBER_OF_FREE_TIERS) {
    const report = ml`
        | You are trying to configure more repositories than there are available in your plan.
        | Update your current plan to access all the features LabelSync offers.
        `
    return { ok: false, error: report }
  }

  /**
   * Report wildcard configuration.
   */
  if (Object.keys(config.repos).includes('*')) {
    const report = ml`
        | You are trying to configure a wildcard configuration on a free plan.
        | Update your current plan to access all the features LabelSync offers.
        `
    return { ok: false, error: report }
  }

  return { ok: true }
}

/* CONTENT CHECK */

/**
 * Returns configuration if everything seems as expected. Throws otherwise
 * with the error exaplanation.
 */
function validateConfiguration(
  config: LSCConfiguration,
): { ok: true; config: LSCConfiguration } | { ok: false; error: string } {
  /**
   * We accumulate possible errors in all the checks and throw the collection.
   */
  let errors: string[] = []

  /* MISSING SIBLINGS */

  /**
   * All siblings should be defined in the configuration.
   */

  let missingSiblings: { repo: string; sibling: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos[repoName]
    const configuredLabels = Object.keys(repoConfig.labels)

    /* Check every label for unconfigured siblings. */
    for (const label in repoConfig.labels) {
      const siblings = repoConfig.labels[label].siblings ?? []

      for (const sibling of siblings) {
        /* Report unconfigured missing sibling. */
        if (!configuredLabels.includes(sibling)) {
          missingSiblings.push({ repo: repoName, sibling, label })
        }
      }
    }
  }

  if (missingSiblings.length !== 0) {
    const report = ml`
    | Configuration references unconfigured siblings:
    | ${missingSiblings.map((s) => `* \`${s.sibling}\` in ${s.repo}:${s.label}`).join(os.EOL)}
    `
    errors.push(report)
  }

  /* ALIAS SPLITING */

  /**
   * You shouldn't make two new labels aliasing one label.
   */

  let duplicateAliaii: { repo: string; alias: string; labels: string[] }[] = []

  /* Check each repository */
  for (const repoName in config.repos) {
    const repoConfig = config.repos[repoName]

    let aliaiiInRepo: { [alias: string]: string[] } = {}

    /* Check every label for unconfigured siblings. */
    for (const label in repoConfig.labels) {
      const aliaii = repoConfig.labels[label].alias ?? []

      /* Create a map of aliases. */
      for (const alias of aliaii) {
        if (!aliaiiInRepo.hasOwnProperty(alias)) aliaiiInRepo[alias] = []
        aliaiiInRepo[alias].push(label)
      }
    }

    /* Find multiples */
    for (const alias in aliaiiInRepo) {
      if (aliaiiInRepo[alias].length > 1) {
        duplicateAliaii.push({
          repo: repoName,
          alias,
          labels: aliaiiInRepo[alias],
        })
      }
    }
  }

  if (duplicateAliaii.length !== 0) {
    const report = ml`
    | Configuration references same old label twice in alias:
    | ${duplicateAliaii.map((s) => `* \`${s.alias}\` appears in ${s.repo}:${s.labels.join(', ')}`).join(os.EOL)}
    |
    | Each legacy label may only be referenced once in alias.
    `
    errors.push(report)
  }

  /* SELF REFERENCES */

  /**
   * Siblings shouldn't self-reference themselves.
   */

  let selfreferencingSiblings: { repo: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos[repoName]

    /**
     * Check every label if it references itself.
     */
    for (const label in repoConfig.labels) {
      const siblings = repoConfig.labels[label].siblings ?? []

      if (siblings.includes(label)) {
        selfreferencingSiblings.push({ repo: repoName, label })
      }
    }
  }

  if (selfreferencingSiblings.length !== 0) {
    const report = ml`
    | Some of the labels in the configuration reference itself as a sibling:
    | ${selfreferencingSiblings.map((s) => `* \`${s.label}\` in ${s.repo}`).join(os.EOL)}
    `
    errors.push(report)
  }

  /* DESCRIPTION LENGTHS */

  /**
   * Siblings shouldn't self-reference themselves.
   */

  let labelsWithTooLongDescriptions: { repo: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos[repoName]

    /**
     * Check every label if it references itself.
     */
    for (const label in repoConfig.labels) {
      const { description } = repoConfig.labels[label]

      if (description && description.length > 1000) {
        labelsWithTooLongDescriptions.push({ repo: repoName, label })
      }
    }
  }

  if (labelsWithTooLongDescriptions.length !== 0) {
    const report = ml`
     | Some of the labels in the configuration have too long descriptions:
     | ${selfreferencingSiblings.map((s) => `* \`${s.label}\` in ${s.repo}`).join(os.EOL)}
      
     | The maximum is 1000 characters.
     `
    errors.push(report)
  }

  /**
   * Collect the errors and throw if necessary.
   */

  if (errors.length > 0) {
    return { ok: false, error: errors.join(os.EOL) }
  }

  return { ok: true, config }
}

/**
 * TRANSFORMATIONS
 */

/**
 * Normalizes configuration that's easier to work with.
 */
function normalizeConfig(config: LSCConfiguration): LSCConfiguration {
  const repos = mapEntries(config.repos, fixRepoConfig)
  const lowercaseRepos = mapKeys(repos, (repo) => repo.toLowerCase())

  return {
    ...config,
    repos: lowercaseRepos,
  }
}

function fixRepoConfig(config: LSCRepository): LSCRepository {
  const labels = mapEntries<LSCLabel, LSCLabel>(config.labels, (label) => ({
    ...label,
    color: normalizeColor(label.color),
  }))

  return {
    ...config,
    labels,
  }
}

export function normalizeColor(color: string): string {
  if (color.startsWith('#')) return color.slice(1).toLowerCase()
  return color.toLowerCase()
}

/**
 * Returns a list of repositories in configuration without wildcard config.
 */
export function getPhysicalRepositories(config: LSCConfiguration): string[] {
  return Object.keys(config.repos).filter((repo) => repo !== '*')
}

/**
 * Tells whether the repository is the configuration
 * repository for the account.
 */
export function isConfigRepo(account: string, repo: string): boolean {
  return getLSConfigRepoName(account) === repo.toLowerCase()
}
