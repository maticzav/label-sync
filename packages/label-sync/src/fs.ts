import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import { Maybe } from './utils'

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export function findFolderUp(dir: string, name: string): Maybe<string> {
  return findUp(dir, (p) => fs.lstatSync(p).isDirectory() && path.basename(p) === name)
}

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export function findFileUp(dir: string, name: string): Maybe<string> {
  return findUp(dir, (p) => fs.lstatSync(p).isFile() && path.basename(p) === name)
}

/**
 * Traverses file system up to find a specific directory that contains
 * a matching pattern.
 *
 * @param dir
 * @param pattern
 */
export function findUp(dir: string, pattern: (path: string) => boolean): Maybe<string> {
  switch (path.normalize(dir)) {
    /* End case: we reached the root. */
    /* istanbul ignore next */
    case '/': {
      return null
    }
    /* Recursive case. */
    default: {
      const elements = fs.readdirSync(dir)
      const includes = elements.map((name) => path.resolve(dir, name)).some(pattern)

      if (includes) return dir
      else return findUp(path.resolve(dir, '../'), pattern)
    }
  }
}
