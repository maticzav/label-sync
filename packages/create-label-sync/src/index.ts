#!/usr/bin/env node

import * as creato from 'creato'
import * as fs from 'fs'
import * as handlebars from 'handlebars'
import * as inquirer from 'inquirer'
import meow from 'meow'
import * as mkdirp from 'mkdirp'
import { ml } from 'multilines'
import ora from 'ora'
import * as os from 'os'
import * as path from 'path'
import * as prettier from 'prettier'

import { loadTreeFromPath, mapEntries, writeTreeToPath } from './utils'

/* Templates */

export const templates: (creato.Template & { identifier: string })[] = [
  {
    identifier: 'yaml',
    name: 'YAML configuraiton',
    description: '',
    repo: {
      branch: 'master',
      path: '/templates/yaml',
      uri: 'https://github.com/maticzav/label-sync',
    },
  },
  {
    identifier: 'typescript',
    name: 'TypeScript configuration',
    description: '',
    repo: {
      branch: 'master',
      path: '/templates/typescript',
      uri: 'https://github.com/maticzav/label-sync',
    },
  },
]

/* Main */

async function main(
  cli: meow.Result<{
    force: {
      alias: string
      type: 'boolean'
      default: false
    }
    template: {
      alias: string
      type: 'string'
    }
    path: {
      alias: string
      type: 'string'
    }
  }>,
) {
  /* Owner information and repository. */

  let owner: string | undefined = undefined

  /* Set a default if we are alrady in repo. */
  if (process.cwd().endsWith('labelsync')) {
    const repository = path.basename(process.cwd())
    owner = repository.replace('-labelsync', '')
  }

  owner = await inquirer
    .prompt<{ owner: string }>([
      {
        name: 'owner',
        message: "What's your personal or organisation GitHub name?",
        type: 'input',
        default: owner,
        validate: (account) => account.trim() !== '',
      },
    ])
    .then((res) => res.owner)

  const repository = `${owner}-labelsync`

  /* Personalised repositories */

  const { repositoriesRaw } = await inquirer.prompt<{
    repositoriesRaw: string
  }>([
    {
      name: 'repositoriesRaw',
      message:
        'Type in repositories that LabelSync should include in configuration (comma-split):',
      type: 'input',
    },
  ])

  const repositories = repositoriesRaw
    .split(',')
    .map((repo) => repo.trim())
    .filter(Boolean)
    .map((name) => ({ name }))

  /* Template */

  let template

  /* Process user input */
  if (cli.flags.template) {
    const matchingTemplate = templates.find(
      (temp) => temp.identifier === cli.flags.template,
    )

    if (!matchingTemplate) {
      const availableTemplates = templates
        .map((temp) => `"${temp.identifier}"`)
        .join(', ')

      console.warn(
        /* prettier-ignore */
        `Couldn't find a template ${cli.flags.template}. You may choose among ${availableTemplates}`,
      )

      process.exit(1)
    }

    template = matchingTemplate
  }

  /* Prompt for templte */
  if (!template) {
    template = await inquirer
      .prompt<{ template: creato.Template }>([
        {
          name: 'template',
          message: 'Choose a LabelSync configuraiton template:',
          type: 'list',
          choices: templates.map((template) => ({
            name: template.name,
            value: template,
          })),
        },
      ])
      .then((res) => res.template)
  }

  /* Check whether we already are in configuration repository. */

  let dist = repository

  if (process.cwd().endsWith('labelsync')) {
    dist = '.'
  }

  if (cli.flags.path) {
    dist = cli.flags.path
  }

  // make dist absolute
  dist = path.resolve(process.cwd(), dist)

  /* Scaffold */
  let approve = cli.flags.force

  /* Prompt for approval if not forced. */
  if (!approve) {
    approve = await inquirer
      .prompt<{ approve: boolean }>([
        {
          name: 'approve',
          message: `We'll scaffold configuration to ${dist}`,
          type: 'confirm',
        },
      ])
      .then((res) => res.approve)
  }

  if (!approve) {
    console.log(`OK! Not scaffolding.`)
    process.exit(0)
  }

  /**
   * Scaffold the template.
   */

  /* Setup tmpdir */
  const tmpDist = path.resolve(os.tmpdir(), `./labelsync`)
  fs.rmdirSync(tmpDist, { recursive: true })
  fs.mkdirSync(tmpDist, { recursive: true })

  if (!fs.existsSync(dist)) mkdirp.sync(dist)

  /* Load template */

  const templateSpinner = ora({
    text: `Downloading ${template.name} template.`,
  }).start()

  const res = await creato.loadTemplate(template, tmpDist)

  if (res.status === 'ok') {
    templateSpinner.succeed()
  } else {
    templateSpinner.fail()
    console.warn(res.message)
    process.exit(1)
  }

  /* Populate */

  const populateSpinner = ora({
    text: `Personalising configuration for you`,
  }).start()

  const tree = loadTreeFromPath(tmpDist, [])

  const populatedTree = mapEntries(tree, (file, name) => {
    try {
      /* Personalise files */
      const populatedFile = handlebars.compile(file)({
        repository,
        repositories,
      })

      /* Format it */
      switch (path.extname(name)) {
        case '.ts': {
          return prettier.format(populatedFile, { parser: 'typescript' })
        }
        case '.yml': {
          return prettier.format(populatedFile, { parser: 'yaml' })
        }
        case '.md': {
          return prettier.format(populatedFile, { parser: 'markdown' })
        }
        case '.json': {
          return prettier.format(populatedFile, { parser: 'json' })
        }
        default: {
          return populatedFile
        }
      }
    } catch (err) {
      templateSpinner.fail()
      console.error(`Error in ${name}`, err.message)
      process.exit(1)
    }
  })
  await writeTreeToPath(dist, populatedTree)

  populateSpinner.succeed(`Template loaded inside ${dist}!`)
}

/* CLI */

const cli = meow(
  ml`
  | create-label-sync
  |
  | > Scaffolds the configuration repostiory for Github Labels manager.
  |
  | Flags:
  |    - force (f): force creates the repository
  |    - template (t): ${templates
    .map((temp) => `"${temp.identifier}"`)
    .join(', ')}
  |    - path (p): folder to scaffold the configuration to 
`,
  {
    flags: {
      force: {
        alias: 'f',
        type: 'boolean',
        default: false,
      },
      template: {
        alias: 't',
        type: 'string',
      },
      path: {
        alias: 'p',
        type: 'string',
      },
    },
  },
)

/* istanbul ignore next */
if (require.main?.filename === __filename) {
  main(cli)
}
