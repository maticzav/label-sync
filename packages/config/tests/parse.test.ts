import fs from 'fs'
import path from 'path'

import { parseConfig, getPhysicalRepositories, isConfigRepo } from '../src/'

type Plan = 'PAID' | 'FREE'

const configurationsPath = path.resolve(__dirname, './__fixtures__/configurations')
const configurations = fs.readdirSync(configurationsPath).map((config) => ({
  config: config,
  path: path.resolve(configurationsPath, config),
}))

describe('configurations:', () => {
  const plans: Plan[] = ['PAID', 'FREE']

  for (const plan of plans) {
    for (const { config, path } of configurations) {
      test(`${config} on ${plan}`, () => {
        const config = parseConfig({
          isPro: plan === 'PAID',
          input: fs.readFileSync(path, { encoding: 'utf-8' }),
        })
        expect(config).toMatchSnapshot()
      })
    }
  }
})

describe('utility functions:', () => {
  test('configRepos returns correct repositories from config', async () => {
    const configPath = path.resolve(__dirname, './__fixtures__/configurations/wildcard.yml')
    const parsedConfig = parseConfig({
      isPro: true,
      input: fs.readFileSync(configPath, { encoding: 'utf-8' }),
    })

    if (!parsedConfig.ok) {
      fail()
      return
    }

    expect(getPhysicalRepositories(parsedConfig.config)).toEqual(['prisma-test-utils'])
  })

  test('isConfigRepo', async () => {
    expect(isConfigRepo('acc', 'acc-labelsync')).toBeTruthy()
    expect(isConfigRepo('ACC', 'acc-labelsync')).toBeTruthy()
    expect(isConfigRepo('acc', 'ACC-labelsync')).toBeTruthy()
    expect(isConfigRepo('ACC', 'ACC-labelsync')).toBeTruthy()

    expect(isConfigRepo('not', 'acc-labelsync')).toBeFalsy()
  })
})
