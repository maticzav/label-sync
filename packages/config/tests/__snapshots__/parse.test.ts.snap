// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configurations: anchors.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {},
  },
  "ok": true,
}
`;

exports[`configurations: anchors.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {},
  },
  "ok": true,
}
`;

exports[`configurations: basic.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "prisma-test-utils": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: basic.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "prisma-test-utils": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: capitalization.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "lowercase": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
      "mixed": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
      "uppercase": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: capitalization.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "lowercase": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
      "mixed": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
      "uppercase": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label": Object {
            "color": "ff0022",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: duplicate_aliases.yml on FREE 1`] = `
Object {
  "error": "Configuration references same old label twice in alias:
* \`alias\` appears in repo-two:label/4, label/5

Each legacy label may only be referenced once in alias.",
  "ok": false,
}
`;

exports[`configurations: duplicate_aliases.yml on PAID 1`] = `
Object {
  "error": "Configuration references same old label twice in alias:
* \`alias\` appears in repo-two:label/4, label/5

Each legacy label may only be referenced once in alias.",
  "ok": false,
}
`;

exports[`configurations: empty.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {},
  },
  "ok": true,
}
`;

exports[`configurations: empty.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {},
  },
  "ok": true,
}
`;

exports[`configurations: free.yml on FREE 1`] = `
Object {
  "error": "You are trying to configure more repositories than there are available in your plan.
Update your current plan to access all the features LabelSync offers.",
  "ok": false,
}
`;

exports[`configurations: free.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "label-sync": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "last-one": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "notquitethere": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "overthetop": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "prisma-test-utils": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "unchanged": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: invalid.yml on FREE 1`] = `
Object {
  "error": "[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"object\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"repos\\"
    ],
    \\"message\\": \\"Required\\"
  }
]",
  "ok": false,
}
`;

exports[`configurations: invalid.yml on PAID 1`] = `
Object {
  "error": "[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"object\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"repos\\"
    ],
    \\"message\\": \\"Required\\"
  }
]",
  "ok": false,
}
`;

exports[`configurations: missing_siblings.yml on FREE 1`] = `
Object {
  "error": "Configuration references unconfigured siblings:
* \`label/6\` in repo-one:label/1
* \`label/7\` in repo-one:label/1
* \`label/8\` in repo-two:label/4",
  "ok": false,
}
`;

exports[`configurations: missing_siblings.yml on PAID 1`] = `
Object {
  "error": "Configuration references unconfigured siblings:
* \`label/6\` in repo-one:label/1
* \`label/7\` in repo-one:label/1
* \`label/8\` in repo-two:label/4",
  "ok": false,
}
`;

exports[`configurations: multiple_aliases.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "repo-one": Object {
        "labels": Object {
          "label/1": Object {
            "alias": Array [
              "alias",
              "two",
            ],
            "color": "ff0022",
          },
          "label/2": Object {
            "color": "ff0022",
          },
          "label/3": Object {
            "color": "ff0022",
          },
        },
      },
      "repo-two": Object {
        "labels": Object {
          "label/4": Object {
            "alias": Array [
              "alias",
            ],
            "color": "ff0022",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: multiple_aliases.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "repo-one": Object {
        "labels": Object {
          "label/1": Object {
            "alias": Array [
              "alias",
              "two",
            ],
            "color": "ff0022",
          },
          "label/2": Object {
            "color": "ff0022",
          },
          "label/3": Object {
            "color": "ff0022",
          },
        },
      },
      "repo-two": Object {
        "labels": Object {
          "label/4": Object {
            "alias": Array [
              "alias",
            ],
            "color": "ff0022",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: new.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "newrepo": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: new.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "newrepo": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: siblings.yml on FREE 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "changed": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label/regular": Object {
            "color": "000000",
          },
          "label/siblings": Object {
            "color": "000000",
            "siblings": Array [
              "label/regular",
            ],
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: siblings.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "changed": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "label/regular": Object {
            "color": "000000",
          },
          "label/siblings": Object {
            "color": "000000",
            "siblings": Array [
              "label/regular",
            ],
          },
        },
      },
    },
  },
  "ok": true,
}
`;

exports[`configurations: wildcard.yml on FREE 1`] = `
Object {
  "error": "You are trying to configure a wildcard configuration on a free plan.
Update your current plan to access all the features LabelSync offers.",
  "ok": false,
}
`;

exports[`configurations: wildcard.yml on PAID 1`] = `
Object {
  "config": Object {
    "repos": Object {
      "*": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
      "prisma-test-utils": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
        },
      },
    },
  },
  "ok": true,
}
`;
